lib/OpenQA/WebAPI/GruJob.pm

```diff
@@ -42,15 +42,22 @@ sub execute {
         $err = $self->SUPER::execute;
     };
 
-    if (defined $err) {
+    # Non-Gru tasks
+    return $err unless $gru_id;
+
+    $info = $self->info;
+    my $state = $info->{state};
+    if ($state eq 'failed' || defined $err) {
+        $err //= $info->{result};
         log_error("Gru command issue: $err");
-        $self->_fail_gru($self->info->{notes}{gru_id} => $err)
-          if $self->fail({(output => $buffer) x !!(defined $buffer), error => $err})
-          && exists $self->info->{notes}{gru_id};
+        $self->fail({defined $buffer ? (output => $buffer) : (), error => $err});
+        $self->_fail_gru($gru_id => $err);
+    }
+    elsif ($state eq 'active') {
+        $self->_delete_gru($gru_id) if $self->finish(defined $buffer ? $buffer : 'Job successfully executed');
```

Comment:
In case you were wondering why this is a separate case from the finished state. Here we avoid a possible race condition where a task calls `$job->retry()` and the retried job gets picked up by a worker fast enough to be in the active state before reaching this line of code. Due to the built-in versioning of Minion jobs the `$self->finish` call here would fail, preventing the Gru task from getting deleted.

Reply:
Now, testing this race condition is hard, and something i'm still thinking about.

Reply:
Maybe it is worth adding this as a comment in the code.

Reply:
Good idea.